plugins/logitech-rallysystem/fu-logitech-rallysystem-audio-device.c

```diff
@@ -0,0 +1,279 @@
+const guint8 kLogiDefaultRallySystemAudioSleepIntervalMs = 100;
+
+struct _FuLogitechRallySystemAudioDevice {
+	FuUdevDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuLogitechRallySystemAudioDevice,
+	      fu_logitech_rallysystem_audio_device,
+	      FU_TYPE_UDEV_DEVICE)
+
+static gboolean
+fu_logitech_rallysystem_audio_device_set_feature(FuLogitechRallySystemAudioDevice *self,
+						 const guint8 *data,
+						 guint datasz,
+						 GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	fu_dump_raw(G_LOG_DOMAIN, "HidSetFeature", data, datasz);
+	return fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				    HIDIOCSFEATURE(datasz),
+				    (guint8 *)data,
+				    NULL,
+				    FU_LOGITECH_RALLYSYSTEM_AUDIO_DEVICE_IOCTL_TIMEOUT,
+				    error);
+#else
+	/* failed */
+	g_set_error_literal(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_SUPPORTED,
+			    "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_logitech_rallysystem_audio_device_get_feature(FuLogitechRallySystemAudioDevice *self,
+						 guint8 *data,
+						 guint datasz,
+						 GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	g_autoptr(GError) error_local = NULL;
+	fu_dump_raw(G_LOG_DOMAIN, "HidGetFeatureReq", data, datasz);
+
+	/* try HIDIOCGINPUT request in case of failure */
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  HIDIOCGFEATURE(datasz),
+				  data,
+				  NULL,
+				  FU_LOGITECH_RALLYSYSTEM_AUDIO_DEVICE_IOCTL_TIMEOUT,
+				  &error_local)) {
+		g_debug("failed to send get request, retrying: %s", error_local->message);
+		if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+					  HIDIOCGINPUT(datasz),
+					  data,
+					  NULL,
+					  FU_LOGITECH_RALLYSYSTEM_AUDIO_DEVICE_IOCTL_TIMEOUT,
+					  error))
+			return FALSE;
+	}
+
+	fu_dump_raw(G_LOG_DOMAIN, "HidGetFeatureRes", data, datasz);
+
+	return TRUE;
+#else
+	/* failed */
+	g_set_error_literal(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_SUPPORTED,
+			    "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_logitech_rallysystem_audio_device_set_version(FuLogitechRallySystemAudioDevice *self,
+						 GError **error)
+{
+	guint8 get_data[TOPOLOGY_DATA_LEN];
+	guint32 fwversion = 0;
+	guint8 index = (kLogiSystemTopologySize * kLogiBolideSysTopologyMaxElements) +
+		       kLogiBolideVersionInfoOffset;
+	/* use temporary variables to avoid [-Wsequence-point] warnings */
+	guint8 major = 0;
+	guint8 minor = 0;
+	guint8 build = 0;
+
+	/* setup HID report to query current device version */
+	get_data[0] = kBolideSystemTopologyHidId;
+	if (!fu_logitech_rallysystem_audio_device_get_feature(self,
+							      (guint8 *)get_data,
+							      TOPOLOGY_DATA_LEN,
+							      error))
+		return FALSE;
+	major = get_data[++index];
+	minor = get_data[++index];
+	build = get_data[++index];
+	/*  little-endian data. MinorVersion byte 0, MajorVersion byte 1, BuildVersion byte 2 */
+	fwversion = (major << 24) + (minor << 16) + build;
```

Comment:
maybe fu_memread_uint24()?

Reply:
Not working for me, need some help here.
fwversion = fu_memread_uint24(&get_data[index+1], G_LITTLE_ENDIAN); // returns 2949377. In place of 16842797

Reply:
isn't that BIG endian?

Reply:
No, it is 3  straight bytes in the get_data buffer (offset index+1): 01 01 2d 
BIG yields: 65837
Both G_BYTE_ORDER & G_LITTLE_ENDIAN yields same: 2949377
![image](https://github.com/fwupd/fwupd/assets/52230102/7b26cfeb-aa7c-45bc-ba68-e341baedcf08)

Reply:
01012d on last line represents version "1.1.45"

Reply:
Sample code and output
	guint8 my_data[5] = {0, 0x01, 0x01, 0x2D, 0}; // 1.1.45 
	guint8 major = my_data[1];
	guint8 minor = my_data[2];
	guint8 build = my_data[3];
	guint32 fwversion = (major << 24) + (minor << 16) + build;
	g_debug("orig test %u", fwversion);
	fwversion = fu_memread_uint24(&my_data[1], G_BYTE_ORDER);
	g_debug("G_BYTE_ORDER %u", fwversion);
	fwversion = fu_memread_uint24(&my_data[1], G_LITTLE_ENDIAN);
	g_debug("G_LITTLE_ENDIAN %u", fwversion);
	fwversion = fu_memread_uint24(&my_data[1], G_BIG_ENDIAN);
	g_debug("G_BIG_ENDIAN %u", fwversion);

20:43:51.958 FuPluginLogitechRallySystem orig test 16842797 /* Correctly Translate into 1.1.45 */
20:43:51.958 FuPluginLogitechRallySystem G_BYTE_ORDER 2949377
20:43:51.958 FuPluginLogitechRallySystem G_LITTLE_ENDIAN 2949377
20:43:51.958 FuPluginLogitechRallySystem G_BIG_ENDIAN 65837

Reply:
Okay, something is broken -- ignore this request and leave it to me. Thanks!

Reply:
Ok, I think I worked it out, fixed.